src/storage/raw/raw_mvcc.rs

```diff
@@ -0,0 +1,215 @@
+        self.seek_first_key_value_cf(Some(cf), Some(opts), key)
+    }
+
+    fn iter(&self, iter_opt: IterOptions) -> Result<Self::Iter> {
+        Ok(RawMvccIterator::new(self.snap.iter(iter_opt)?))
+    }
+
+    fn iter_cf(&self, cf: CfName, iter_opt: IterOptions) -> Result<Self::Iter> {
+        Ok(RawMvccIterator::new(self.snap.iter_cf(cf, iter_opt)?))
+    }
+
+    #[inline]
+    fn lower_bound(&self) -> Option<&[u8]> {
+        self.snap.lower_bound()
+    }
+
+    #[inline]
+    fn upper_bound(&self) -> Option<&[u8]> {
+        self.snap.upper_bound()
+    }
+
+    fn ext(&self) -> S::Ext<'_> {
+        self.snap.ext()
+    }
+}
+
+pub struct RawMvccIterator<I: Iterator> {
+    inner: I,
+    cur_key: Option<Vec<u8>>,
+    cur_value: Option<Vec<u8>>,
+    is_valid: Option<bool>,
+}
+
+fn is_user_key_eq(left: &[u8], right: &[u8]) -> bool {
+    let len = left.len();
+    if len != right.len() {
+        return false;
+    }
+    Key::is_user_key_eq(left, &right[..len - U64_SIZE])
+}
+
+impl<I: Iterator> RawMvccIterator<I> {
+    fn new(inner: I) -> Self {
+        RawMvccIterator {
+            inner,
+            cur_key: None,
+            cur_value: None,
+            is_valid: None,
+        }
+    }
+
+    fn update_cur_kv(&mut self) {
+        if let Some(ref mut key) = self.cur_key {
+            key.clear();
+            key.extend_from_slice(self.inner.key());
+        } else {
+            self.cur_key = Some(Vec::from(self.inner.key()));
+        };
+        if let Some(ref mut value) = self.cur_value {
+            value.clear();
+            value.extend_from_slice(self.inner.value());
+        } else {
+            self.cur_value = Some(Vec::from(self.inner.value()));
+        };
+        self.is_valid = Some(true);
+    }
+
+    fn clear_cur_kv(&mut self) {
+        self.cur_key = None;
+        self.cur_value = None;
+        self.is_valid = None;
+    }
+
+    fn move_to_prev_max_ts(&mut self) -> Result<bool> {
+        if self.inner.valid()? {
+            self.update_cur_kv();
+            self.inner.prev()?;
+        } else {
+            self.clear_cur_kv();
+            return Ok(false);
+        }
+        while self.inner.valid()? {
+            // cur_key should not be None here.
+            if is_user_key_eq(self.cur_key.as_ref().unwrap(), self.inner.key()) {
+                self.update_cur_kv();
+                self.inner.prev()?;
+            } else {
+                break;
+            }
+        }
+        Ok(true)
+    }
+}
+
+impl<I: Iterator> Iterator for RawMvccIterator<I> {
+    fn next(&mut self) -> Result<bool> {
+        let cur_key = self.inner.key().to_owned();
+        let mut res = self.inner.next()?;
+        while res && self.inner.valid()? && is_user_key_eq(&cur_key, self.inner.key()) {
+            res = self.inner.next()?;
+        }
+        self.clear_cur_kv();
+        Ok(res)
+    }
+
+    fn prev(&mut self) -> Result<bool> {
+        self.move_to_prev_max_ts()
```

Comment:
This is wrong. For example, if there are only one key in the db and iterator is pointing at the only one key, calling prev should return false.

Reply:
In this case, move_to_prev_max_ts actually return Ok(false). when caller call seek_for_prev, move_to_prev_max_ts  is also called once, so engine iterator has been invalid. Then call move_to_prev_max_ts  again will return Ok(false)

Reply:
What if it calls `seek` and then calls `prev`?

Reply:
also works. Caller should check valid() after call prev(), and valid() return false in this case.

Reply:
Is this the proper usage of a iterator? This iterator is exposed to RawStoreInner and it does not have the usage like this.

Reply:
It's not the optimal usage, but It's OK to do so.

Reply:
currently return error if caller call `seek` then `prev` or `seek_for_prev` then `next`